From: Yu Watanabe <watanabe.yu+github@gmail.com>
Date: Thu Nov 7 03:43:50 2024 +0900
Subject: network: keep all dynamically acquired configurations when KeepConfiguration=dhcp-on-stop

By the previous commit, configuration source of addresses and routes are
saved on stop and restored on start. Hence, we can keep dynamic
configurations on stop.

Co-authored-by: Jian Zhang <zhangjian.3032@bytedance.com>

Bug-Ubuntu: https://launchpad.net/bugs/2098183
Origin: backport, https://github.com/systemd/systemd/commit/4eca221ab8d1ce3b95596aef5b56e784d18d41cc
[chengen - fix the context in the link_stop_engines function.]

--- systemd-255.4.orig/src/network/networkd-link.c
+++ systemd-255.4/src/network/networkd-link.c
@@ -337,6 +337,24 @@ int link_stop_engines(Link *link, bool m
                 k = sd_dhcp_client_stop(link->dhcp_client);
                 if (k < 0)
                         r = log_link_warning_errno(link, k, "Could not stop DHCPv4 client: %m");
+
+                k = sd_ipv4ll_stop(link->ipv4ll);
+                if (k < 0)
+                        r = log_link_warning_errno(link, k, "Could not stop IPv4 link-local: %m");
+
+                k = sd_dhcp6_client_stop(link->dhcp6_client);
+                if (k < 0)
+                        r = log_link_warning_errno(link, k, "Could not stop DHCPv6 client: %m");
+
+                k = dhcp_pd_remove(link, /* only_marked = */ false);
+                if (k < 0)
+                        r = log_link_warning_errno(link, k, "Could not remove DHCPv6 PD addresses and routes: %m");
+
+                k = ndisc_stop(link);
+                if (k < 0)
+                        r = log_link_warning_errno(link, k, "Could not stop IPv6 Router Discovery: %m");
+
+                ndisc_flush(link);
         }
 
         k = sd_dhcp_server_stop(link->dhcp_server);
@@ -351,28 +369,10 @@ int link_stop_engines(Link *link, bool m
         if (k < 0)
                 r = log_link_warning_errno(link, k, "Could not stop LLDP Tx: %m");
 
-        k = sd_ipv4ll_stop(link->ipv4ll);
-        if (k < 0)
-                r = log_link_warning_errno(link, k, "Could not stop IPv4 link-local: %m");
-
         k = ipv4acd_stop(link);
         if (k < 0)
                 r = log_link_warning_errno(link, k, "Could not stop IPv4 ACD client: %m");
 
-        k = sd_dhcp6_client_stop(link->dhcp6_client);
-        if (k < 0)
-                r = log_link_warning_errno(link, k, "Could not stop DHCPv6 client: %m");
-
-        k = dhcp_pd_remove(link, /* only_marked = */ false);
-        if (k < 0)
-                r = log_link_warning_errno(link, k, "Could not remove DHCPv6 PD addresses and routes: %m");
-
-        k = ndisc_stop(link);
-        if (k < 0)
-                r = log_link_warning_errno(link, k, "Could not stop IPv6 Router Discovery: %m");
-
-        ndisc_flush(link);
-
         k = sd_radv_stop(link->radv);
         if (k < 0)
                 r = log_link_warning_errno(link, k, "Could not stop IPv6 Router Advertisement: %m");
